File: src/main/java/com/sleepkqq/sololeveling/service/auth/TgAuthService.java
package com.sleepkqq.sololeveling.service.auth;

import static com.sleepkqq.sololeveling.localization.LocalizationMessage.ERROR_AUTH_HASH;
import static com.sleepkqq.sololeveling.localization.LocalizationMessage.ERROR_AUTH_REQUIRED;
import static java.util.Objects.nonNull;
import static org.springframework.security.authentication.UsernamePasswordAuthenticationToken.authenticated;
import static org.springframework.security.web.context.HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY;

import com.sleepkqq.sololeveling.api.UserApi;
import com.sleepkqq.sololeveling.localization.LocalizationException;
import com.sleepkqq.sololeveling.model.TgAuthData;
import com.sleepkqq.sololeveling.model.UserData;
import com.sleepkqq.sololeveling.view.home.HomeView;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.server.VaadinSession;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class TgAuthService {

  private final UserApi userApi;
  private final TgHashService tgHashService;

  public void authenticate(TgAuthData tgAuthData) {
    if (isLoggedIn()) {
      UI.getCurrent().navigate(HomeView.class);
      return;
    }

    var tgWebAppData = tgAuthData.tgWebAppData();
    if (!tgHashService.checkHash(tgAuthData.initData(), tgWebAppData.hash())) {
      log.error("Invalid hash, received data: '{}'", tgWebAppData);
      throw new LocalizationException(ERROR_AUTH_HASH);
    }

    var userData = UserData.fromTgUser(tgWebAppData.user());
    userApi.saveUserInfo(userData);

    var authentication = authenticated(userData, tgWebAppData.hash(), userData.getAuthorities());

    setAuthentication(authentication);
    UI.getCurrent().getPage().reload();

    log.info("User '{}' authenticated successfully", userData.getUsername());
  }

  public void logout() {
    setAuthentication(null);
    UI.getCurrent().getPage().reload();
  }

  public boolean isLoggedIn() {
    return nonNull(getAuthentication());
  }

  public Optional<UserData> findCurrentUser() {
    return Optional.ofNullable(getAuthentication())
        .map(Authentication::getPrincipal)
        .filter(UserData.class::isInstance)
        .map(UserData.class::cast);
  }

  public UserData getCurrentUser() {
    return findCurrentUser().orElseThrow(() -> new LocalizationException(ERROR_AUTH_REQUIRED));
  }

  private static SecurityContext getContext() {
    return SecurityContextHolder.getContext();
  }

  private static Authentication getAuthentication() {
    return getContext().getAuthentication();
  }

  private static void setAuthentication(Authentication authentication) {
    var context = getContext();
    context.setAuthentication(authentication);
    VaadinSession.getCurrent().getSession().setAttribute(SPRING_SECURITY_CONTEXT_KEY, context);
  }
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/service/TgAuthService.kt
package com.sleepkqq.sololeveling.ui.service

import com.sleepkqq.sololeveling.ui.api.UserApi
import com.sleepkqq.sololeveling.ui.localization.LocalizationException
import com.sleepkqq.sololeveling.ui.localization.LocalizationMessage
import com.sleepkqq.sololeveling.ui.model.TgAuthData
import com.sleepkqq.sololeveling.ui.model.UserData
import com.sleepkqq.sololeveling.ui.view.home.HomeView
import com.vaadin.flow.component.UI
import com.vaadin.flow.server.VaadinSession
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContext
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.context.HttpSessionSecurityContextRepository
import org.springframework.stereotype.Service

@Service
class TgAuthService(
	private val userApi: UserApi,
	private val tgHashService: TgHashService
) {

	private val log = LoggerFactory.getLogger(TgAuthService::class.java)

	fun authenticate(tgAuthData: TgAuthData) {
		if (this.isLoggedIn) {
			UI.getCurrent().navigate(HomeView::class.java)
			return
		}

		val tgWebAppData = tgAuthData.tgWebAppData
		if (!tgHashService.checkHash(tgAuthData.initData, tgWebAppData.hash)) {
			log.error("Invalid hash, received data: '{}'", tgAuthData)
			throw LocalizationException(LocalizationMessage.ERROR_AUTH_HASH)
		}

		val userData = UserData.fromTgUser(tgWebAppData.user)
		userApi.authUser(userData)

		authentication = UsernamePasswordAuthenticationToken.authenticated(
			userData,
			tgWebAppData.hash,
			userData.authorities
		)
		UI.getCurrent().page.reload()

		log.info("User '{}' authenticated successfully", userData.username)
	}

	fun logout() {
		authentication = null
		UI.getCurrent().page.reload()
	}

	val isLoggedIn: Boolean
		get() = authentication != null

	fun findCurrentUser(): UserData? {
		return authentication
			?.principal
			?.takeIf { it is UserData }
			?.let { it as UserData }
	}

	val currentUser: UserData
		get() = findCurrentUser()
			?: throw LocalizationException(LocalizationMessage.ERROR_AUTH_REQUIRED)

	private var authentication: Authentication?
		get() = context.authentication
		set(authentication) {
			context.authentication = authentication
			VaadinSession.getCurrent()
				.session
				.setAttribute(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY, context)
		}

	private val context: SecurityContext
		get() = SecurityContextHolder.getContext()
}